### ISSUE TITLE
Transcoding broken no port defined

### ISSUE THREAD

**Issue by lemmingsbeest** [E2OpenPlugins/e2openplugin-OpenWebif#247]
When you click the icon in webif for transcode your ip adress is shown in vlc but no port number..

It shows this line:
http://192.168.1.24:None/1:0:16:1793:C82:3:EB0000:0:0:0:

As you see it writes None behind the ip adress....

**Comment by Schimmelreiter** [comment]
Which image/distro?

**Comment by lemmingsbeest** [comment]
Box is Vu Duo 2
Image is openpli 4.0 1-2-2015

**Comment by Schimmelreiter** [comment]
That's exactly what I expected.

OpenPLi refused to implement the transcoding in a way compatible with
- Vu+ original transcoding
- the oe-a variant of Vu+ transcoding
or
- the Xtrend/Mut@nt way of dynamic transcoding.

Even more:
They implemented it in a way its presence can not reliably be detected (Abusing file names already in use for other purposes on other images).

So none of the port/param detections inside OpenWebif currently works with their extrawurst way of transcoding.

If OpenPLi wants working transcoding again, they have to either fix their broken transstreamproxy AND/OR to provide a proper patch to OpenWebif to allow OWIF to read their extrawurst config.

The easiest way for them would be to make Erik Slagter's Vu+-transstreamproxy compatible with the transcoding on Xtrend 10000/Mut@nt, as that is what it is currently closest to, so they probably won't.

**Comment by WanWizard** [comment]
You don't support Xtrend/Mut@ant either, so why refer to that?

VU+ transcoding sucks, it has a separate binary and a separate plugin, just to set some parameters. Which not all work, are instable, and only allow you to set them globally. OpenPLi allows you to set transcoding parameters per stream. So what exactly is "extrawurst"?

I can not recall you asking for anything "easily detectable". Instead of bashing why don't you do something constructive so that end-users can enjoy transcoding on all hardware and images that support it?

**Comment by Schimmelreiter** [comment]
OWIF itself supports transcoding on any box that supports it and it even does it in a way so that it works on oe-a images as well as Vu+/VTI ones with only a single .py differing.

The Vu+ transcoding has the benefit that it actually works. Erik's transcoding might have theoretical advantages but they are outweighted by practical problems:
Users have asked him for a transcoding setup plugin and for a long time he refused to re-implement that. Yes, his transcoding CAN be configured dynamically but so can the one for Mut@nt and Xtrend 10000 ... but the later ones manage to do so without making the configuration unavailable from E2 and it manages to provide a source for default values if a client does NOT send all those dynamic configuration params.

I've asked Erik not to name his binary /usr/bin/streamproxy, which already is the file name for the untranscoded streaming proxy so it can not be used to detect the presence of PLi transcoding.
But he refused.

I've asked for methods to determine the capabilities and the presence of transcoding support and its port and all he answered was to query the transstreamproxy via http query which requires the asking party to already know before what we actually want to find out ... the port ...

And I'm not bashing, I'm only stating facts. For each of the points above you can find a post inside the broken transcoding megathread where I've mentioned my considerations and asked for more cooperative solutions.
Even now I have pointed out an easy solution to make transcoding work instantly on PLI:
If you copy the Mut@nt/Xtrend dynamic transcoding, just copy it completely and provide the settings in the same way as Mut@nt and Xtrend do and OWIF will automagically read them rather than providing yet another island solution.

If OpenPLi continues on its way to behave totally uncooperative, OpenPLi will also have to provide the necessary patches to keep things working on OpenPLi.

OpenPLi can now either finally return to a compatible behaviour or provide patches as above.

As long as the patches don't break compatibility both is perfectly fine for me.

All I'm saying:
I'm not going to spend MY time on OpenPLi workarounds but OpenPLi devs are welcome to waste theirs.

BTW: extrawurst = special sausage.
"Someone always requiring an extrawurst" describes persons that always refuse to go with the program. E.g. if all people in a chat room agree on using UFT-8, the one person that still insists on using MacRoman is the one who always needs some extrawurst.

**Comment by WanWizard** [comment]
I don't really understand this.

I happen to have hardware from all three vendors supporting transcoding, and only on the Duo2 OWIF gives gives me a second streaming icon for a transcoded stream. Which doesn't work, but it used to. An OpenPLi image uses the same URL and the same config file as any other image. The plugin you refer to only updates the config file values, it doesn't do ANYTHING related to transcoding. And the resulting config is there. So that entire discussion is not relevant.

I don't know where and how you have the Xtrend and the Mut@nt working, I haven't seen it. A quick glance at the code reveals no support for these brands either, only VU+ hardware is referenced. For Xtrend, I noticed a separate issue (#248) is opened, so I'm not the only one who has noticed it. By the way, we haven't copied anything, we have had a big hand into the way Xtrend and Mut@nt handle transcoding, because we didn't want them to copy the crap VU+ produced.

If you create software that must support multiple platforms and multiple vendors, you are bound to have to deal with that. Differences between images will only become bigger over time. The architecture of Enigma is crap, and the way everyone deals with the different platforms doesn't really help either, simply because there is no unified way of dealing with features.

This is why we are pushing the BSP layer, and why we want all vendor and hardware specific stuff out of the base image. The BSP layer should dictate the feature set, and should contain all vendor specific items.

This applies to OWIF as well: it should not be your job to figure out which features are available, by using unreliable methods like file or proc value checks. Enigma should have a unified interface to tell you the features, and this should be provided from the vendors BSP layer. Same with stuff like for example remote controls, for which OWIF keeps it's own versions, it's own detection mechanism, etc.

You keep moaning about the fact that we want to do things differently. We don't. We want to do things better. And that requires doing things differently. And we believe that these are improvements that will benefit everyone. But that requires the will to change. The more people want to push this change, the quicker we have a complete BSP layer and we all can create true platform independent code. And that is a good thing.

**Comment by Schimmelreiter** [comment]
See https://github.com/E2OpenPlugins/e2openplugin-OpenWebif/blob/master/plugin/controllers/models/stream.py beginning at line 55.

The first block handles the non-dynamic original Vu+, Gigablue and Edision transcoding, that's also where the port detection fails on OpenPLi because it doesn't use that kind of transcoding and it doesn't have the config.plugins.transcodingsetup.port value.

I would have happily added an exception to the if clause there to make OpenPLi not jump into this block (because it doesn't work for OpenPLi) if there was a simply way to detect OpenPLi (Joining the oe-a and/or providing a proper own implementation of boxbranding would have helped here).

The second block handles the dynamic transcoding on Mut@nt and et10000.
In theory this would be the block OpenPLi would have to use for Vu+ boxes too, but OpenPLi doesn't use port 8001 for the transcoding like Mut@nt and Xtrend do
and OpenPLi doesn't provide the default values for transcoded streams where this block expects them ...

So:
OpenPLi would either need to make their transcoding compatible to one of those existing methods (Preferably the later one), then all we would need is a proper condition for the if clauses to make OpenPLi use the Mut@nt way for Vu+ too
or
it needs its completely own block which has not to break compatibility with the other blocks.

That's it.

**Comment by WanWizard** [comment]
Ok. so if I understand it correctly: if that config key is added to the settings (manually if needed), it should start working on the Duo2? That seems to be a pretty easy fix, the plugin can write those. I'll take that back for discussion.

And for the second block: I happen to have an ET10K in front of me now, but no transcoding icon to be found in its OWIF. So if you say that should work (since we haven't done anything special for transcoding on those boxes), I say it doesn't, so more work is needed for these boxes.

I see here (https://github.com/E2OpenPlugins/e2openplugin-OpenWebif/blob/master/plugin/controllers/transcoding.py#L30) the same config values are checked, so I assume adding them would also enable the icon on the 2400HD and the 10K?

**Comment by Schimmelreiter** [comment]
Providing the setting
config.plugins.transcodingsetup.port
(setting it to 8002 or whatever eriks transstreamproxy is set to use)
would be the quick fix.

But actually eriks transstreamproxy is more like xtrend's, so modifying the if clause "if machinebuild in ('inihdp', 'hd2400', 'et10000'):" in line 68 to also match for Vu+ (and the if clause in line 55 NOT to match Vu+) on OpenPLi would enable to use that way of transcoding.
However, that variant of transcoding uses one port (8001) for everything, normal and transcoded streaming and it reads the defaults for transcoded streaming from /etc/enigma2/settings rather than /etc/streamproxy.conf or whatever erik called it.

In my opinion, if eriks streamproxy resembles the way xtrends transcoder works, it should also behave fully compatible to it, i.e. not use port 8002 but 8001 (eriks transspreamproxy can handle untranscoded streaming as well).

For the transcoding icon missing on OpenPLi there might be various reasons.
The simpliest would be the if clause in
https://github.com/E2OpenPlugins/e2openplugin-OpenWebif/blob/master/plugin/controllers/models/stream.py#L68 not matching.
This problem would originate inside owibranding.py, the module that TRIES to provide the same info as the branding module on oe-a.
It tries to return functional equivalent return values but not 100% identical ones.


As you said it yourself:
"This applies to OWIF as well: it should not be your job to figure out which features are available, by using unreliable methods like file or proc value checks. Enigma should have a unified interface to tell you the features, and this should be provided from the vendors BSP layer."
oe-a's branding module is a huge step into this direction and the owibranding.py is just an ATTEMPT to return the same but it is still based on all those proc readings.
Basically it just moves all of the unreliable detection into a single py so that the rest of OWIF can become more box independent, but the detections remain a PITA.
oe-a's branding module sets the values at build time and provides the proper remote control, box image, ... automagically.

Or to point it out:
Theoretically OpenPLi is on the wrong fork of OWIF. The e2open master is enough for old OE1.6 images like VTI and Black Hole that are available for only a handful of boxes and it works fairly ok for those boxes that have no advanced feature sets like old Dream boxes.

The more boxes you support and the more advanced they are, the more you will see that the way of adding more and more detections, remotes, box pictures and so on to this plugin becomes a dead end road.

**Comment by WanWizard** [comment]
Ok, I'll try to find the time (and to be honest some Python skills) to do some debugging and give you feedback on it. I'll talk to Erik to see what we can do in terms of the VU+.

As to OE-A's branding, as said before we don't think that is the way forward. We believe the vendor or manufacturer should be in charge and deliver everything that is hardware and/or model specific, which is why we introduced the BSP layer. We are not there yet, currently is contains only code, but our aim is to expand it to everything that is currrently handled by OE-A's branding, handling both compile time and runtime box specific information.

**Comment by macnuts** [comment]
OpenPLi can be fixed quickly (for vu+ family, not sure how other boxes are detected) in aforementioned **stream.py** file by replacing:
```python
59			#Transcoding Plugin is not installed or your STB does not support transcoding
60			transcoder_port = None
```
with:
```python
			if info.getImageVersion() == "4" and info.getImageDistro() == "openpli" :
				transcoder_port = 8002
			else:
				#Transcoding Plugin is not installed or your STB does not support transcoding
				transcoder_port = None
```
The same in **getTS()** routine maybe (for recordings?). You could use **8003** instead of **8002** there.

A step further would be replacing 8002 and 8003 values with the values parsed from **/etc/enigma2/streamproxy.conf**:
>listen = 8002:transcode
>listen = 8003:transcode
>size = 480p
>bitrate = 500
>profile = baseline
>level = 3.1
>bframes = 0

but that would brake your **owibranding.py** effort, I guess.

**Comment by Schimmelreiter** [comment]
I've considered the first approach but decided against:
1. The "imagedistro" detection is the most unreliable one in owibranding. Its that bad that I even thought about not implementing it at all. It's o.k. for the cosmetical use (In Box Info) but should not be relied upon until OpenPLi adds a proper branding module.
2. It would work for "openpli" for the moment, but there are images based on openpli but that aren't openpli. Considering the reliability of the detection, owibranding would need to get bloated with more and more of those unreliable detections to also handle openrsi and so on.
3. transcoder_port = 8002 would still be an assumption, as it can be changed. Pre-declaring it as "None" and doing proper reading of its actual value from config serves a purpose later (Enables it to fail if it is still "None" after all detections)

Reading "/etc/enigma2/streamproxy.conf" actually would be an option but that's up to OpenPLi to add it. If I did, I couldn't test it.
It's not really clean, as you can have "/etc/enigma2/streamproxy.conf" but no transcoding support at all, as packages do not remove their conf files on uninstall, but it would work to some degree.

As I pointed out before:
It would be the easiest thing to just stop being different without being better by simply putting the transcoding settings into Enigma2's genuine config.
Nothing prevents OpenPLi from keeping their very own transcoding, all I ask for is to store its config in a compatible way.
In this case, their Vu+ transcoding settings should be compatible to other image's Xtrend transcoding, as their Vu+ transcoder behaves exactly that way.

If OpenPLi did just that, they would get a transcoding setup plugin for free (Xtrend's) and working transcoding that can even benefit from Erik's dynamic transcoding features without ANY changes in OWIF.

Saying that, if we have a win-win option on the one hand and an ugly workaround on the other, I'm not going to move any finger for the workaround.

**Comment by satsis** [comment]
How fix problem with transcoding on VU+ Duo2? OpenPli
programmers openpli said that are not responsible for this plug-in, and nothing helped.

http://192.168.0.11:None/1:0:19:4E86:14:70:1680000:0:0:0:'.

root@vuduo2:~# opkg remove enigma2-plugin-extensions-openwebif
Removing package enigma2-plugin-extensions-openwebif from root...
root@vuduo2:~# opkg install enigma2-plugin-extensions-openwebif
Installing enigma2-plugin-extensions-openwebif (1+git1284+515760d-r0.3) on root.
Downloading http://downloads.pli...r0.3_vuduo2.ipk.
Configuring enigma2-plugin-extensions-openwebif.
root@vuduo2:~#

**Comment by Schimmelreiter** [comment]
PLi made their very own transcoder for Vu+ rather than using genuine transcoding.
So it's entirely out of question that PLi als has to provide the fix for their trancoding.

Until they have done that, you can try about any other distro, which will work just fine (Except if it's just an OpenPLi rip-off).
OpenATV, OpenViX, OpenHDF, OpenSpa, VTI, Black Hole, .... will work.

**Comment by WanWizard** [comment]
Transcoding is completely compatible. Problem is that OpenWebIf has an external dependency (to the config file created by the VU+ transcoding plugin), but doesn't deal with this dependency. 

It just assumes it exists instead of making sure all dependencies are provided. Very poor programming practice, and very cheap to shift the blame to someone else.

**Comment by satsis** [comment]
The essence of the problem is that it is a link in the plugin OpenWebif not formed correctly, ie instead of port 8002 gives none.
If you use an application dreamdroid, port 8002 work fine.
It is a pity that no one wants fix problem on the web interface

http://forums.openpli.org/topic/38259-problem-with-transciding-and-update-image/

**Comment by WanWizard** [comment]
It is most likely because OpenWebIf assumes that the transcoding config file, that defines the port, exists, while on OpenPLi it doesn't by default. 

It is a dependency of the transcoding plugin, so if you install that and start it, it will be created. And it is a dependency of OpenWebIf, but they don't want to provide this dependency as they should, which is why you have this problem. They can't even bother to display an error message if the config file that seems to be required is not present.

**Comment by satsis** [comment]
They also ask for the money to be experts.
It is a pity that there are no guilty.
It is a pity that no one correct.

**Comment by Schimmelreiter** [comment]
Sorry, there is nothing I can do for you.
As pointed out, the problem is within the PLi image and it either has to be fixed there or at least specifically for it, which requires a developer using OpenPLi.

I would try to fix it, but as this problem is not the only one which OpenPLi considers as "none of our businesses" I have stopped using OpenPLi in favor of OpenATV long ago, so I can't.

**Comment by satsis** [comment]
http://forums.openpli.org/topic/38259-problem-with-transciding-and-update-image/page-2#entry497262

(((

**Comment by phenex80** [comment]
Hi guys!

I'm used the snippet from [macnuts](#issuecomment-73268986) which is worked __perfectly__
Thanks!
G.

**Comment by eriksl** [comment]
It's not necessary. Transcoding on all receivers supported by OpenPLi works ALSO over the OpenWebIF.

If people are to stubborn to follow our instructions (install the transcoding setup plugin, which probably already is installed), and run it once, then I can't take the blame for that. I've spend enough time to make it all compliant with the silly interface between image en OpenWebIF.

The flaw in the reasoning was and still is that you'd want to "configure" a transcoder. At OpenPLi we have always stayed far from that. A transcoder should not be configured. A transcoded stream should be requested and only at the moment the stream is available and the request can be satisfied, then the transconder (the one that is assigned, which you never know beforehand) should be configured.

And that is why we will never supply a plugin that does that, which OpenWebIF expects. So I made a plugin that emulates this behaviour to satisfy OWIF but I am going to regret it more and more, if people don't even want to use it.

**Comment by phenex80** [comment]
Hi Eric!

	I’m truly sorry! I’m a newbie in linux STBs. I’m bought a VU+ STB at this week, after VTI, BH I’m tried your OpenPLI and I’m stay because everything works automatically fine, the usage is like in other commercial STBs, but the transcoded streaming isn’t work for me and I can’t found instruction how to set it up.

	I’ll be absollutely hapy if you could send me a link for the instruction for the correct setup.
	With this „dirty” changes VU+ app from AppStor works too.

	Thank you for your time and your response.

	Have a nice day.

	Gabriel.

**Comment by eriksl** [comment]
Like I said, run the "stub" plugin transcoding setup once, that will create the environment the OWIF expects and then it works.

**Comment by phenex80** [comment]
Hi Erik! 
I'm tried your instructions. From Webif I could play the stream, but the m3u for the transcoded version of the stream is still w/o port:
```
#EXTM3U 
#EXTVLCOPT--http-reconnect=true 
#EXTINF:-1,Discovery Science
http://vu.lan:None/1:0:1:77F2:2C3:600:E080000:0:0:0:
```
Also the VU+ iPad application stop works, but never mind the VLC works with this setup.

**Comment by eriksl** [comment]
You may have to restart the receiver first.

**Comment by phenex80** [comment]
Hi Erik!
I'm install the updates for webif and transcoding at today, but the problem with port :None at my Vu+ Solo SE V2 still remain.
I followed all of your instructions, but nothing. :/
THX.
G.

**Comment by eriksl** [comment]
Did you install and run the transcoding setup plugin?

**Comment by phenex80** [comment]
I've installed _enigma2-plugin-systemplugins-transcodingsetup - experimental-git4035+726c8fa-r0.3_ and setupped as bitrate 1000000 and  resolution 480p (I read it from Webif now).

**Comment by eriksl** [comment]
And then restart the receiver once.

**Comment by phenex80** [comment]
Sure, several times, because I intalled it when I bought the STB, cca. a month ago.

**Comment by eriksl** [comment]
Open an issue on the forum.

**Comment by phenex80** [comment]
Do you mean at http://forums.openpli.org?

**Comment by eriksl** [comment]
Yes.

**Comment by phenex80** [comment]
OK. Thanks
